### PR TITLE
feat/commonjs-incremental

### PR DESCRIPTION
**Pull request by marionebl**

### CODE REVIEW

File: source/application/tasks/commonjs/index.js [hunk 1]
```diff
@@ -1,17 +1,177 @@
...
+			return true;
+		}
+
+		// Get the types in this pattern
+		const types = [...new Set(pattern.files
+			.map(path => extname(path).slice(1))
+			.filter(Boolean)
+			.map(key => patterns.formats[key])
+			.filter(Boolean)
+			.map(config => config.name))];
```
Comment (KnisterPeter) [inline source/application/tasks/commonjs/index.js]: I would recommend to not name this parameter config, but format which is what you expect from the map function two lines above. The wordings in this statement are not that easy to get

File: source/application/tasks/commonjs/index.js [hunk 2]
```diff
@@ -1,17 +1,177 @@
...
+		// Build if pattern mtime > artifact mtime
+		if (pattern.mtime.getTime() > artifact.mtime.getTime()) {
+			return true;
+		}
+
+		// Get the types in this pattern
+		const types = [...new Set(pattern.files
+			.map(path => extname(path).slice(1))
+			.filter(Boolean)
+			.map(key => patterns.formats[key])
```
Comment (KnisterPeter) [inline source/application/tasks/commonjs/index.js]: The key parameter should be named extensions, I think

File: source/library/utilities/get-pattern-mtimes.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+function getLatestMtime(mtimes) {
+	const times = mtimes.map(mtime => {
+		return {stamp: mtime.getTime(), date: mtime}
+	});
+
+	const latest = times.sort((a, b) => b.stamp - a.stamp)[0];
+	return latest.date;
+}
+
+function getDependencyMtimes(pattern, patterns) {
```
Comment (KnisterPeter) [inline source/library/utilities/get-pattern-mtimes.js:34]: Okay for now I think, but this should belong to a global cache so all tasks and transforms can access this and do not have to do file-io all the time.

**Comment by KnisterPeter** [comment]
Awesome. Besides the few comments it is very clean and readable :+1:

File: source/library/utilities/get-pattern-mtimes.js (hunk 3 above)
Comment (marionebl) [inline source/library/utilities/get-pattern-mtimes.js:34]: I'm with you on this, stating and deriving subtrees to build probably should be the central optimization technique in patternplate. Thinking about the best way to land this in `patternplate-core`

File: source/application/tasks/commonjs/index.js (hunk 1 above)
Comment (marionebl) [inline source/application/tasks/commonjs/index.js]: Done

File: source/application/tasks/commonjs/index.js (hunk 2 above)
Comment (marionebl) [inline source/application/tasks/commonjs/index.js]: Done

**Pull request merged**